Userland/xargs.cpp

```diff
@@ -0,0 +1,201 @@
+            }
+            parts.append(arg.substring_view(previous_index + replace_view.length(), arg.length() - previous_index - replace_view.length()));
+        } else {
+            parts.append(arg);
+        }
+        argument_parts.append(move(parts));
+        ++index;
+    }
+
+    // Append a '{}' at the end if no arguments have replacements.
+    if (!some_argument_has_replacement) {
+        Vector<StringView> parts;
+        parts.append("");
+        parts.append("");
+        argument_parts.append(move(parts));
+    }
+
+    FILE* fp = stdin;
+    bool is_stdin = true;
+
+    if (StringView { "-" } != file_to_read) {
+        // A file was specified, try to open it.
+        fp = fopen(file_to_read, "r");
+        if (!fp) {
+            perror("fopen");
+            return 1;
+        }
+        is_stdin = false;
+    }
+
+    char* line = nullptr;
+    size_t buffer_size = 0;
+
+    StringBuilder builder;
+    Vector<char*> child_argv;
+
+    child_argv.resize(argument_parts.size() + 2);
+    child_argv[0] = const_cast<char*>(command);
+    child_argv[child_argv.size() - 1] = 0;
+
+    int devnull_fd = 0;
+
+    if (is_stdin) {
+        devnull_fd = open("/dev/null", O_RDONLY | O_CLOEXEC);
+        if (devnull_fd < 0) {
+            perror("open");
+            return 1;
+        }
+    }
+
+    bool fail = false;
+
+    for (;;) {
+        auto line_size = getdelim(&line, &buffer_size, entry_separator, fp);
+
+        if (line_size < 0) {
+            perror("getdelim");
+            fail = true;
+            break;
+        }
+
+        if (line_size == 0)
+            break;
+
+        if (line[line_size - 1] == '\n')
+            --line_size;
+
+        StringView line_view { line, (size_t)line_size };
+
+        size_t index = 0;
+        for (auto& parts : argument_parts) {
+            builder.clear();
+            builder.join(line_view, parts);
+            child_argv[index + 1] = strdup(builder.build().characters());
+            ++index;
+        }
+
+        if (verbose) {
+            fprintf(stderr, "xargs: ");
+            for (auto* ptr : child_argv) {
+                if (!ptr)
+                    break;
+                fprintf(stderr, "%s ", ptr);
+            }
+            fprintf(stderr, "\n");
+            fflush(stderr);
+        }
+
+        auto pid = fork();
+        if (pid < 0) {
+            perror("fork");
+            return 1;
+        }
+
+        if (pid == 0) {
+            if (is_stdin)
+                dup2(devnull_fd, STDIN_FILENO);
+
+            execvp(command, child_argv.data());
+
+        } else {
+            int wstatus = 0;
+            if (waitpid(pid, &wstatus, 0) < 0) {
+                perror("waitpid");
+                return 1;
+            }
+            // FIXME: Should we take a look at the exit code and fail if the command fails?
+            for (size_t i = 0; i < argument_parts.size(); ++i)
+                free(child_argv[i + 1]);
+
+            free(line);
```

Comment:
Why wait for completion before freeing these all?

Reply:
No particular reason, I guess it just felt better this way.
I'll move it up, I guess